snowpack/src/sources/local.ts

```diff
@@ -96,22 +119,147 @@ type PackageImportData = {
   packageVersion: string;
   packageName: string;
 };
-const allPackageImports: Record<string, PackageImportData> = {};
-const allSymlinkImports: Record<string, string> = {};
-const allKnownSpecs = new Set<string>();
-const inProgressBuilds = new PQueue({concurrency: 1});
-let hasWorkspaceWarningFired = false;
-
-export function getLinkedUrl(builtUrl: string) {
-  return allSymlinkImports[builtUrl];
-}
 
-/**
- * Local Package Source: A generic interface through which Snowpack
- * interacts with esinstall and your locally installed dependencies.
- */
-export default {
-  async load(id: string, isSSR: boolean) {
+export class PackageSourceLocal implements PackageSource {
+  config: SnowpackConfig;
+  arb: Arborist;
+  packageRootDirectory: string;
+  lockfile: LockfileManifest = {dependencies: {}, lock: {}};
+  memoizedResolve: Record<string, string> = {};
+  allPackageImports: Record<string, PackageImportData> = {};
+  allSymlinkImports: Record<string, string> = {};
+  allKnownSpecs = new Set<string>();
+  allKnownProjectSpecs = new Set<string>();
+  hasWorkspaceWarningFired = false;
+
+  constructor(config: SnowpackConfig) {
+    this.config = config;
+    this.arb = new Arborist({
+      ...(typeof config.packageOptions.source === 'string' ? {} : config.packageOptions.source),
+      path: PKG_SOURCE_DIR,
+      packageLockOnly: true,
+    });
+    this.packageRootDirectory =
+      config.packageOptions.source === 'local' ? config.root : PKG_SOURCE_DIR;
+  }
+
+  private async installPackageRootDirectory(installTargets: InstallTarget[]) {
+    const {arb, config} = this;
+    const result = await arb.loadVirtual().catch(() => null);
+    const packageNamesNeedInstall = new Set(
+      installTargets
+        .map((spec) => {
+          const specParts = spec.specifier.split('/');
+          let _packageName: string = specParts.shift()!;
+          if (_packageName?.startsWith('@')) {
+            _packageName += '/' + specParts.shift();
+          }
+          // IMPRVE DOCS: handle aliases
+          const aliasEntry = findMatchingAliasEntry(config, _packageName);
+          if (aliasEntry && aliasEntry.type === 'package') {
+            const {from, to} = aliasEntry;
+            _packageName = _packageName.replace(from, to);
+          }
+          if (!this.lockfile.dependencies[_packageName]) {
+            return _packageName;
+          }
+          // Needed to make TS happy. Gets filtered out in next step.
```

Comment:
Yeah TS has trouble with `.filter()` because it‘s not very statically-analyzable.